kotlin/samples/dungeon/timemachine/src/main/java/com/squareup/sample/timemachine/TimeMachineWorkflow.kt

```diff
@@ -0,0 +1,90 @@
+import com.squareup.workflow.WorkflowAction
+import com.squareup.workflow.WorkflowAction.Mutator
+import com.squareup.workflow.renderChild
+import kotlin.time.Clock
+import kotlin.time.Duration
+import kotlin.time.ExperimentalTime
+
+/**
+ * A workflow that will continuously render [delegateWorkflow] and either record its renderings, or
+ * play them back via the [rendering][TimeMachineRendering] depending on the mode selected by this
+ * workflow's [props][TimeMachineProps].
+ *
+ * Any outputs from the [delegateWorkflow] will be forwarded as-is, even in playback mode.
+ *
+ * @param delegateWorkflow The [Workflow] whose renderings to record. This workflow will be rendered
+ * continuously as long as the `TimeMachineWorkflow` is being rendered.
+ * @param clock The [Clock] to use to assign timestamps to recorded values.
+ */
+@ExperimentalTime
+class TimeMachineWorkflow<P, out O : Any, out R>(
+  private val delegateWorkflow: Workflow<P, O, R>,
+  clock: Clock
+) : StatelessWorkflow<TimeMachineProps<P>, O, TimeMachineRendering<R>>() {
+
+  /**
+   * Sealed class that determines whether a [TimeMachineWorkflow] is [recording][Recording] the
+   * [delegate workflow's][delegateWorkflow] renderings, or [playing them back][PlayingBackAt].
+   */
+  sealed class TimeMachineProps<out P> {
+    /**
+     * The props to pass to the [delegateWorkflow].
+     */
+    abstract val delegateProps: P
+
+    /**
+     * Puts the [TimeMachineWorkflow] in recording mode.
+     *
+     * The rendering returned by the [delegateWorkflow] will be recorded.
+     *
+     * @param delegateProps The props to pass to the [delegateWorkflow].
+     */
+    data class Recording<out P>(override val delegateProps: P) : TimeMachineProps<P>()
+
+    /**
+     * Puts the [TimeMachineWorkflow] in playback mode.
+     *
+     * The [delegateWorkflow] will still be rendered, but that rendering will be dropped.
+     *
+     * @param delegateProps The props to pass to the [delegateWorkflow].
+     * @param timestamp The timestamp nearest to the recorded value to return in the
+     * [TimeMachineRendering].
+     */
+    data class PlayingBackAt<out P>(
+      override val delegateProps: P,
+      val timestamp: Duration
+    ) : TimeMachineProps<P>()
+  }
+
+  private val recordingWorkflow = RecorderWorkflow<R>(clock)
+
+  override fun render(
+    props: TimeMachineProps<P>,
+    context: RenderContext<Nothing, O>
+  ): TimeMachineRendering<R> {
+    // Always render the delegate, even if in playback mode, to keep it alive.
+    val delegateRendering =
+      context.renderChild(delegateWorkflow, props.delegateProps) { forwardOutput(it) }
+
+    val recorderProps = when (props) {
+      is Recording -> RecordValue(delegateRendering)
+      is PlayingBackAt -> PlaybackAt(props.timestamp)
+    }
+
+    return context.renderChild(recordingWorkflow, recorderProps)
+  }
+
+  private fun forwardOutput(output: O) = object : WorkflowAction<Nothing, O> {
```

Comment:
off topic - is there a reason we only have a `workflowAction {}` extension for `StatelessWorkflow` ?